domain/src/main/kotlin/com/krzykrucz/elesson/currentlesson/preparedness/domain/DomainAPI.kt

```diff
@@ -0,0 +1,75 @@
+import com.krzykrucz.elesson.currentlesson.preparedness.domain.UnpreparednessError.AlreadyRaised
+import com.krzykrucz.elesson.currentlesson.preparedness.domain.UnpreparednessError.LessonNotStarted
+import com.krzykrucz.elesson.currentlesson.preparedness.domain.UnpreparednessError.StudentNotPresent
+import com.krzykrucz.elesson.currentlesson.preparedness.readmodel.StudentInSemesterReadError
+import com.krzykrucz.elesson.currentlesson.preparedness.readmodel.StudentSubjectUnpreparednessInASemester
+import com.krzykrucz.elesson.currentlesson.shared.AsyncOutput
+import com.krzykrucz.elesson.currentlesson.shared.ClassName
+import com.krzykrucz.elesson.currentlesson.shared.CurrentLesson
+import com.krzykrucz.elesson.currentlesson.shared.FirstName
+import com.krzykrucz.elesson.currentlesson.shared.LessonIdentifier
+import com.krzykrucz.elesson.currentlesson.shared.Output
+import com.krzykrucz.elesson.currentlesson.shared.SecondName
+import java.time.LocalDateTime
+
+//types
+data class UnpreparedStudent(
+        val firstName: FirstName,
+        val secondName: SecondName
+)
+
+data class StudentsUnpreparedForLesson(val students: List<UnpreparedStudent> = emptyList())
+
+sealed class UnpreparednessError {
+    object AlreadyRaised : UnpreparednessError()
+    object UnpreparedTooManyTimes : UnpreparednessError()
+    object TooLateToRaiseUnpreparedness : UnpreparednessError()
+    object StudentNotPresent : UnpreparednessError()
+    object LessonNotStarted: UnpreparednessError()
+}
+
+data class StudentReportingUnpreparedness(
+        val firstName: String,
+        val secondName: String
+)
+
+//dependency
+typealias CheckNumberOfTimesStudentWasUnpreparedInSemester = (PresentStudent, ClassName) -> AsyncOutput<StudentSubjectUnpreparednessInASemester, StudentInSemesterReadError>
+
+typealias HasStudentUsedAllUnpreparednesses = (StudentSubjectUnpreparednessInASemester) -> Boolean
+
+typealias HasStudentAlreadyRaisedUnprepared = (StudentsUnpreparedForLesson, PresentStudent) -> Boolean
+
+typealias CheckStudentIsPresent = (StudentReportingUnpreparedness, CheckedAttendanceList) -> Output<PresentStudent, StudentNotPresent>
+
+typealias AreStudentsEqual = (PresentStudent, StudentReportingUnpreparedness) -> Boolean
+
+//persistence
+typealias PersistUnpreparedStudentToLesson = (StudentMarkedUnprepared) -> IO<LessonIdentifier>
+
+typealias NotifyStudentMarkedUnprepared = (StudentMarkedUnprepared) -> IO<Unit>
+
+typealias FindCurrentLesson = (LessonIdentifier) -> AsyncOutput<CurrentLesson, LessonNotStarted>
+
+//workflows
+typealias CheckStudentCanReportUnprepared = (PresentStudent, ClassName) -> AsyncOutput<PresentStudent, UnpreparednessError>
+
+typealias NoteStudentUnpreparedForLesson = (PresentStudent, StudentsUnpreparedForLesson) -> Output<StudentsUnpreparedForLesson, AlreadyRaised>
+
+typealias CreateEvent = (LessonIdentifier, StudentsUnpreparedForLesson) -> StudentMarkedUnprepared
```

Comment:
it's intriguing that `notification/persistence etc` are defined in the domain but not used in any workflow defined here but only in the implementation, shouldn't it be defined on the implementation layer